tlsn/tlsn-verifier/src/tls/mod.rs

```diff
@@ -0,0 +1,439 @@
+
+use mpz_core::serialize::CanonicalSerialize;
+use mpz_garble::{config::Role as GarbleRole, protocol::deap::DEAPVm};
+use mpz_ot::{
+    actor::kos::{
+        msgs::Message as ActorMessage, ReceiverActor, SenderActor, SharedReceiver, SharedSender,
+    },
+    chou_orlandi, kos,
+};
+use mpz_share_conversion as ff;
+use mpz_share_conversion::ShareConversionVerify;
+use rand::Rng;
+use signature::Signer;
+use tls_mpc::{setup_components, MpcTlsFollower, TlsRole};
+use tlsn_core::{
+    msg::{SignedSessionHeader, TlsnMessage},
+    HandshakeSummary, SessionHeader, Signature,
+};
+use uid_mux::{yamux, UidYamux};
+use utils_aio::{codec::BincodeMux, duplex::Duplex, expect_msg_or_err, mux::MuxChannel};
+
+use crate::Mux;
+
+#[cfg(feature = "tracing")]
+use tracing::{debug, info, instrument};
+
+type OTSenderActor = SenderActor<
+    chou_orlandi::Receiver,
+    SplitSink<
+        Box<dyn Duplex<ActorMessage<chou_orlandi::msgs::Message>>>,
+        ActorMessage<chou_orlandi::msgs::Message>,
+    >,
+    SplitStream<Box<dyn Duplex<ActorMessage<chou_orlandi::msgs::Message>>>>,
+>;
+
+/// A Verifier instance.
+pub struct Verifier<T: state::VerifierState> {
+    config: VerifierConfig,
+    state: T,
+}
+
+impl Verifier<state::Initialized> {
+    /// Create a new verifier.
+    pub fn new(config: VerifierConfig) -> Self {
+        Self {
+            config,
+            state: state::Initialized,
+        }
+    }
+
+    /// Set up the verifier.
+    ///
+    /// This performs all MPC setup.
+    ///
+    /// # Arguments
+    ///
+    /// * `socket` - The socket to the prover.
+    pub async fn setup<S: AsyncWrite + AsyncRead + Send + Unpin + 'static>(
+        self,
+        socket: S,
+    ) -> Result<Verifier<state::Setup>, VerifierError> {
+        let mut mux = UidYamux::new(yamux::Config::default(), socket, yamux::Mode::Server);
+        let mux_control = BincodeMux::new(mux.control());
+
+        let mut mux_fut = MuxFuture {
+            fut: Box::pin(async move { mux.run().await.map_err(VerifierError::from) }.fuse()),
+        };
+
+        let encoder_seed: [u8; 32] = rand::rngs::OsRng.gen();
+        let mpc_setup_fut = setup_mpc_backend(&self.config, mux_control.clone(), encoder_seed);
+        let (mpc_tls, vm, ot_send, ot_recv, gf2, ot_fut) = futures::select! {
+            res = mpc_setup_fut.fuse() => res?,
+            _ = &mut mux_fut => return Err(std::io::Error::from(std::io::ErrorKind::UnexpectedEof))?,
+        };
+
+        Ok(Verifier {
+            config: self.config,
+            state: state::Setup {
+                mux: mux_control,
+                mux_fut,
+                mpc_tls,
+                vm,
+                ot_send,
+                ot_recv,
+                ot_fut,
+                gf2,
+                encoder_seed,
+            },
+        })
+    }
+
+    /// Runs the TLS verifier to completion, notarizing the TLS session.
+    pub async fn notarize<S: AsyncWrite + AsyncRead + Send + Unpin + 'static, T>(
```

Comment:
Can we rename this function into `execute_full_notarization`, or something better and shorter :) which captures the fact that it just runs the whole notary workflow?

Reply:
Maybe `setup_and_notarize`?

Reply:
Because there is another `notarize` function, and I think this would communicate the intent more precisely.